### ISSUE TITLE
namespaces don't work properly when defined within functions

### ISSUE THREAD

**Issue by stephane-chazelas** [ksh93/ksh#325]
From https://unix.stackexchange.com/questions/668176/problem-with-ksh-namespaces-defined-inside-functions

```
$ function f { a=0; namespace x { a=1; }; typeset -p .x a; }; f
namespace x
{
        :
}
a=1
```

Expected:

```
$ function f { a=0; namespace x { a=1; }; typeset -p .x a; }; f
namespace x
{
        a=1
}
a=0
```

See  also:

```
$ namespace x { a=1; }; .x.b=2; typeset -p .x b
namespace x
{
        a=1
}
b=2
```

Expected:

```
$ namespace x { a=1; }; .x.b=2; typeset -p .x b
namespace x
{
        a=1
        b=2
}
```

I see namespaces were apparently removed in ksh2020, maybe they were deemed non-functional?

**Comment by stephane-chazelas** [comment]
> I see namespaces were apparently removed in ksh2020, maybe they were deemed non-functional?

Ah, sorry, namespaces are still there in ksh2020, it's just using `namespace` inside a function seems to cause a syntax error:

```
$ ksh2020 -c 'function f { namespace x { a=1; }; }'
ksh: syntax error at line 1: `namespace' unexpected
```

Though you can do:

```
$ ksh2020 -c 'function f { eval "namespace x { a=1; }"; typeset -p .x; }; f'
namespace x
{
        :
}
```

Which exhibits the same bug.

**Comment by hyenias** [comment]
@stephane-chazelas I see several things going on here as described by your examples. Here are my notes:

> $ function f { a=0; namespace x { a=1; }; typeset -p .x a; }; f

I was not aware that global namespaces could be created inside the new style ksh functions--apparently they can. But it seems you have exposed some bugs. First,  the hierarchy/placement/namespace information is being removed/lost/altered for the 2nd variable creation/assignment of 'a'; this is evidenced by having variable `a`'s value changed to 1. If I use the old style global functions aka `f()` it works as you expected.

```
$ ksh -c 'function f { a=0; namespace x { a=1; }; typeset -p .x a; }; f'
namespace x
{
	:
}
a=1

$ ksh -c 'f() { a=0; namespace x { a=1; }; typeset -p .x a; }; f'
namespace x
{
	a=1
}
a=0
```
Next, if I utilize ksh's new style functions to restrict variable **a** to function scope only the namespace x's **a** variable is not created either.
```
$ ksh -c 'function f { a=0; namespace x { a=1; }; }; f; typeset -p .x a .x.a;'
namespace x
{
	:
}
a=1
.x.a=1
$ ksh -c 'function f { typeset a=0; namespace x { a=1; }; }; f; typeset -p .x a .x.a;'
namespace x
{
	:
}
```
Also note that namespace x's variables are being created when no name collision exists however namespace x's variables are not being displayed properly. This is seen when I altered .x.a to .x.b.

```
$ ksh -c 'function f { a=0; namespace x { b=1; }; }; f; typeset -p .x a .x.b;'
namespace x
{
	:
}
a=0
.x.b=1
$ ksh -c 'function f { typeset a=0; namespace x { b=1; }; }; f; typeset -p .x a .x.b;'
namespace x
{
	:
}
.x.b=1
```

**Comment by hyenias** [comment]
@stephane-chazelas now to comment on your second bug example, `namespace x { a=1; }; .x.b=2; typeset -p .x b`:
```
$ ksh -c 'namespace x { a=1; }; .x.b=2; typeset -p .x b'
namespace x
{
	a=1
}
b=2
```
In my experience with ksh, I have to create a parent variable within a namespace before it can be referenced/altered/unset. I am not allowed to create a parent variable indirectly using the dot notation of a variable assignment. Maybe this functionality can be fixed/incorporated into ksh for variable assignments. See below the use of namespace to create the parent variable. I just use the term parent variable to mean a scalar or container variable. With your example, when the `.x.b=2` assignment occurs the .x namespace is dropped/skipped and that variable **b** gets created into the global namespace.

However, if you define namespace x's **b** variable then it works. I prefer the use of `b=''` or `b=` instead of `typeset b` as I like seeing the initialize variable b--perhaps this is another bug you have uncovered. 
```
$ ksh -c 'namespace x { a=1; b=; }; typeset -p .x; .x.b=2; typeset -p .x b'
namespace x
{
	a=1
	b=''
}
namespace x
{
	a=1
	b=2
}
$ ksh -c 'namespace x { a=1; typeset b; }; typeset -p .x; .x.b=2; typeset -p .x b'
namespace x
{
	a=1
}
namespace x
{
	a=1
	b=2
}
```

**Comment by hyenias** [comment]
Not sure ksh2020 souce code is going to be any help in respects to namespaces as its namespace code was diminished. It seems that the ksh93v- authors decided to not allow namespace commands inside function definitions on 2013-05-29--not to say what direction ksh93u+m will go but bringing this knowledge into the light.

However, I skimmed over ksh93v- beta release notes and they made lots of fixes and changes to namespaces. Here are just a few of the entries from https://github.com/att/ast/blob/2016-01-10-beta/src/cmd/ksh93/RELEASE.

> 13-08-19  A bug in which compound variables created in a namespace did not display correctly outside of the namespace has been fixed.
> 13-08-11 +namespace was modified so that namespace names can be a compound variable rather than just an identifier and namespace are no longer nested.
> 13-07-08  Fixed a namespace bug in which print -v did not work for printing a compound array of a namespace variable.
> **13-05-29  namespace commands are no longer allowed inside function definitions and now generate a syntax error.**
> 12-04-02  Made some namespace changes and added a regression test.

**Comment by hyenias** [comment]
FYI: Played around with ksh2020 to see what a nested namespace would look like. Not sure the value of having nested namespaces as compared to the potential slow down that might be imposed on all variable lookups and such.
```
$ ksh -c 'namespace x { namespace x.y { b=2; }; a=1; }; typeset -p .x .x.y'
ksh: x.y: is not an identifier
$ ksh2020 -c 'namespace x { namespace x.y { b=2; }; a=1; }; typeset -p .x .x.y'
namespace x
{
	a=1
	namespace y
	{
		b=2
	}
}
namespace x.y
{
	b=2
}
```

**Comment by hyenias** [comment]
Here are the referenced additional ksh93v- regression tests for namespaces.

https://github.com/att/ast/blob/e770c77e9816e156c6df4a455e71b5f9fff79310/src/cmd/ksh93/tests/namespace.sh#L106-L187

**Comment by McDutchie** [comment]
> However, I skimmed over ksh93v- beta release notes and they made lots of fixes and changes to namespaces. […]
> > **13-05-29  namespace commands are no longer allowed inside function definitions and now generate a syntax error.**

Wow. What good are namespaces then? The main use case I can think of is having private name spaces for library functions. AFAICT, this is just about as good as getting rid of them entirely.

**Comment by McDutchie** [comment]
I think we should go with the 93v- change and make it an error to define a namespace within a function. It doesn't seem to make much sense to do so; what is useful is defining functions within namespaces, not the other way around.

As an illustration, I played around with namespaces a bit and made the following function that integrates ksh's internal `somecommand --man` documentation with the regular `man` command so you don't have to remember which is which. It uses four subfunctions that are defined within a `man` namespace so their names don't interfere with other functions with the same name. Put the code below in a file in your `$FPATH` to use. As you can see, namespaces allow for a very readable way of structuring your code.

```sh
#!/martijns/man/for/ksh

namespace man
{
	isbuiltin()
	{
		typeset LF=$'\n'
		(($# == 1)) &&
		[[ ! $1 =~ ^(-.*|:|true|false|echo)$ ]] &&
		[[ $LF$(builtin)$LF == *"$LF$(whence "$1")$LF"* ]]
	}

	showbuiltin()
	(
		export ERROR_OPTIONS=emphasis
		command "$1" --man 2>&1
	)

	pager()
	(
		unset IFS
		set -o noglob
		${PAGER:-less -R}
	)

	main()
	{
		# special-case "man 1 foo" == "man foo"
		(($# == 2)) && [[ $1 == "1" ]] && shift

		if	isbuiltin "$@"
		then	showbuiltin "$1" | pager
		else	command man "$@"
		fi
	}
}

function man
{
	.man.main "$@"
}
```

**Comment by McDutchie** [comment]
> FYI: Played around with ksh2020 to see what a nested namespace would look like. 

That looks quite broken in ksh2020. You have to define the nested namespace with `namespace x.y` but `typeset -p` outputs it as `namespace y` which will not work for reinput.

I tend to agree with you that nested namespaces probably don't really have value.

I'm inclined to block defining namespaces within function definitions as well as attempts to define namespaces within namespaces, seeing what other bugfixes can be backported from 93v-, and otherwise leaving well enough alone.

**Comment by hyenias** [comment]
I would like to bring to your attention that the **namespace** keyword can be used for more than just the initial creation/defining of a new namespace. It can be used to **_work with_** existing namespaces. Within a namespace context/code block, you have access to that local namespace/scope. In your previous unedited example, you had to use the fully qualified references all starting with `.man.`. I am not saying that encasing your provided example within a namespace man { ... } code block can remove all of the `.man.`'s, but by using namespace code blocks I can avoid all/most of those dot+name+dot prefixes. This cuts down on errors, less typing, easier to edit, and may be faster as there is less code to parse.  

To me, a function is a collection of code and being able to issue namespace code blocks is beneficial whether that code is issued at the default/global namespace or within a function.

Side note 1: Looking forward to trying out your above man function code. Interesting...
Side note 2: What happens if you source some code within a namespace? I need to give that a try and find out what happens.

**Comment by McDutchie** [comment]
> I would like to bring to your attention that the namespace keyword can be used for more than just the initial creation/defining of a new namespace. It can be used to ***work with*** existing namespaces. 

That is a good point, I had not considered that.

So the `namespace` keyword is not just for defining namespaces, it's also for using them.

Still, I might counter-argue that the correct way of doing that is to move all your code that uses that particular namespace to a function defined within that namespace (which is what I did in the edited man function).

**Comment by hyenias** [comment]
Sure, you can argue that type of coding style. Another way of doing it is to use an alias to cross reference into that namespace (perhaps you do not need the top-level man function). But why limit or force that onto all others? Its just one way of coding. I personally like using namespace code blocks because I tend to always miss that starting dot and when I have to change the name of the namespace or move code from one block to another it inherits the name and easier to alter. All good things to me whereas removing that functionality seems to provide no advantages.

ksh93v- starting enabling the nested namespaces which lead to all sort of bugs--fixed and outstanding. Also, as I stated before I do not see the need for nested namespaces. I can always use an underscore in the namespace name to create similar namespaces for grouping purposes. When ksh93v- started allowing the dot in the name that tends to confuse me as to what might be a compound variable.

**Comment by McDutchie** [comment]
> Sure, you can argue that type of coding style. Another way of doing it is to use an alias to cross reference into that namespace (perhaps you do not need the top-level man function). But why limit or force that onto all others? 

TBH, mainly because it's broken if you don't observe that limitation. Which I imagine is probably the reason why 93v- enforced it. So, either we do better than AT&T and fix it somehow, or we enforce the limitation.

If it takes a long time to fix, then we should enforce the limitation for the next release, until we can fix it.

**Comment by hyenias** [comment]
Understood. Limiting namespace statements within the new style functions is fine with me as fixing it will take a long time. Hopefully, the old style functions will continue to be able to issue namespace code blocks.

**Comment by McDutchie** [comment]
Of course that would still not do anything for the second bug @stephane-chazelas reported at the top. As long as that cannot be fixed, then assigning to namespace variables from outside the namespace (as in `.foo.bar=baz` after `namespace foo`) should be blocked, too.

How about this: these things get blocked in the 1.0 branch only, as the next release will be from that. (And I want to have the next release out Real Soon Now™ – even Slackware got so desperate for a new ksh they've been packaging our development code in slackware-current!) We'll leave it alone in the master branch so it can hopefully be fixed sometime.

**Comment by hyenias** [comment]
Sounds good to me.

**Comment by McDutchie** [comment]
The following seems to effectively prohibit namespace blocks within ksh functions and not in POSIX functions. The check is done at runtime and not (as in 93v-) at parse time.
```diff
diff --git a/src/cmd/ksh93/sh/xec.c b/src/cmd/ksh93/sh/xec.c
index c61d3edc..90bf484d 100644
--- a/src/cmd/ksh93/sh/xec.c
+++ b/src/cmd/ksh93/sh/xec.c
@@ -859,6 +859,11 @@ static Namval_t *enter_namespace(Shell_t *shp, Namval_t *nsp)
 	Namval_t	*path=nsp, *fpath=nsp, *onsp=shp->namespace;
 	Dt_t		*root=0,*oroot=0;
 	char		*val;
+	if(sh.st.real_fun)
+	{
+		error(ERROR_exit(3),"namespaces cannot be defined in a ksh function scope");
+		UNREACHABLE();
+	}
 	if(nsp)
 	{
 		if(!nv_istable(nsp))
```

**Comment by hyenias** [comment]
Unfortunately, that patch did not prevent the namespace x from being created by the new style ksh function. I found a bug with unset leaving behind an empty string for the namespace that required me to perform 2 unsets to cleanup from a previous run. See below:
```
{
echo '------ clean up -----'
unset -f test1 test2
functions
unset .x .y
# if namespaces x and y exist, after unsetting them; ksh is leaving behind .x='' and .y='' variables. another bug.
unset .x .y
typeset -p .x .y
echo '------ cleaned ------'
echo '------ begin   ------'
function test1 { namespace x { a=1; }; }
test2() { namespace y { b=2; }; }
functions
echo '---------------------'
# namespaces x and y should not exist at this time
typeset -p .x .y
echo test1: Testing new style ksh function
test1
typeset -p .x
echo test2: Testing old style posix functions that do not have scope
test2
typeset -p .y
echo '------ end     ------'
}
```
```
$ arch/*/bin/ksh
# pasted in snippet from above interactively
------ clean up -----
------ cleaned ------
------ begin   ------
function test1 { namespace x { a=1; }; }
test2() { namespace y { b=2; }; }
---------------------
test1: Testing new style ksh function
arch/linux.i386-64/bin/ksh[18]: test1: namespaces cannot be defined in a ksh function scope
namespace x
{
	:
}
test2: Testing old style posix functions that do not have scope
namespace y
{
	b=2
}
------ end     ------
$
```

**Comment by McDutchie** [comment]
Right, that makes sense. Try this one instead. Your test script now does not output a `namespace x`.
```diff
--- a/src/cmd/ksh93/sh/xec.c
+++ b/src/cmd/ksh93/sh/xec.c
@@ -2481,6 +2481,11 @@ int sh_exec(register const Shnode_t *t, int flags)
 					errormsg(SH_DICT,ERROR_exit(1),e_ident,fname);
 					UNREACHABLE();
 				}
+				if(shp->st.real_fun)
+				{
+					error(ERROR_exit(3),"namespaces cannot be defined in a ksh function scope");
+					UNREACHABLE();
+				}
 				sfputc(stkp,'.');
 				sfputr(stkp,fname,0);
 				np = nv_open(stkptr(stkp,offset),shp->var_tree,flags);
```

**Comment by McDutchie** [comment]
I think this feature should probably be redesigned mostly from scratch in the development branch. Keep the code that parses the `namespace` keyword and executes those blocks, rip out everything else, and start over.

My reason for this is that the current approach is not only buggy, but complicated. And it should not be complicated, because namespaces, as documented in the manual page, are a very simple idea. They automatically prefix the current namespace name, preceded and followed by a dot, to all variable and function names that are created within the namespace. And when referencing existing variables and functions that don't start with a dot, they search for the prefixed name first, before defaulting to the name without the prefix (if any). That's it.

This should be done in the functions where their names are looked up and stored. For any other purpose, variables and functions within a namespace should remain normal variables and functions. We don't need pretty 'typeset -p' output. Scoping is irrelevant and shouldn't enter into any of this. Namespaces that don't touch or know about scoping won't have any scoping bugs.

**Comment by hyenias** [comment]
Confirmed. Your last patch works as described thus disabling namespace instructions within new style ksh functions.

I certainly want less bugs and if doing so requires a fresh approach then I agree with you. I do find it troublesome that the only way I am able to examine entire namespace code blocks is by parsing out `typeset -p` output. `functions` works to that limited degree. It would be nice to have a new builtin that deals with namespaces in a more effective and consolidated manner. Having reworked namespaces would allow external ksh libraries to be created, imported, and managed better.

**Comment by McDutchie** [comment]
I'm working on the new approach now. Removing the old stuff took some doing – you'd think everything is between `SHOPT_NAMESPACE` tags, but no, of course that code had little tentacles in various other places. But that's done now, and I've started working on the new code and I'm making surprisingly quick progress. It's not quite half working yet, but getting there. It'll be much smaller, simpler and cleaner than the old way. And it'll support nesting, because with the new approach, that's trivial.

There will be no hooks into typeset at all. ksh already has the necessary functionality and we don't need to duplicate it or make it fancier. To print out a namespace 'foo' you'll be able to do `typeset -p "${!.foo.@}"` to get the names starting with `.foo.`.

With what I have now, variable expansions work, assignments don't yet. But this already works:
```
$ arch/*/bin/ksh -xc 'namespace sh { echo $version; }'
arch/darwin.i386-64/bin/ksh: warning: [DEBUG] entering namespace .sh
+ echo Version AJM 93u+m/1.1.0-alpha+dev 2021-11-24
Version AJM 93u+m/1.1.0-alpha+dev 2021-11-24
arch/darwin.i386-64/bin/ksh: warning: [DEBUG] exiting namespace .sh
```

…and that doesn't even work on any released ksh version, though the manual says/implies it should!

**Comment by McDutchie** [comment]
The redesign from scratch is way more difficult than I expected and is not going to happen any time soon, as there are still so many other higher-priority issues to work on. For now I'll just block namespaces within functions on the dev branch the same way I did on the 1.0 branch back in 2021 (40de1e92b07200d0330e66cf1fa9dacc5ab2c99e). This will close this issue. The namespace redesign is going on my long-term to do list, and I may or may not ever get around to it.
